File: convert-price-lists-to-js.py
#!/usr/bin/env python3
"""
Konvertiert die von fill_price_lists_paprika.py erzeugten CSV-Dateien
aus data/price_lists/ in einzelne JavaScript-Dateien unter src/scripts/.

Verwendung:
    python csv_to_js.py
"""

from __future__ import annotations

import csv
import json
import math
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


BASE_DIR = Path(__file__).resolve().parent
INPUT_DIR = BASE_DIR / "data" / "price_lists"
OUTPUT_DIR = BASE_DIR / "src" / "scripts"

# False:
# Preise bleiben Strings mit deutschem Dezimalkomma:
# "63915,48"
#
# True:
# Preisfelder werden echte JavaScript-Zahlen:
# 63915.48
#
# Für Berechnungen in JavaScript ist True normalerweise besser.
PRICE_VALUES_AS_NUMBERS = True

FILES = [
    {
        "input": "bitcoin_prices.csv",
        "output": "price-data-btc.js",
        "variable": "btcPriceData",
        "label": "Bitcoin",
    },
    {
        "input": "gomining_prices.csv",
        "output": "price-data-gmt.js",
        "variable": "gmtPriceData",
        "label": "GoMining",
    },
    {
        "input": "ethereum_prices.csv",
        "output": "price-data-eth.js",
        "variable": "ethPriceData",
        "label": "Ethereum",
    },
    {
        "input": "bnb_prices.csv",
        "output": "price-data-bnb.js",
        "variable": "bnbPriceData",
        "label": "BNB",
    },
    {
        "input": "solana_prices.csv",
        "output": "price-data-sol.js",
        "variable": "solPriceData",
        "label": "Solana",
    },
    {
        "input": "toncoin_prices.csv",
        "output": "price-data-ton.js",
        "variable": "tonPriceData",
        "label": "Toncoin",
    },
    {
        "input": "usdt_prices.csv",
        "output": "price-data-usdt.js",
        "variable": "usdtPriceData",
        "label": "Tether",
    },
    {
        "input": "usdc_prices.csv",
        "output": "price-data-usdc.js",
        "variable": "usdcPriceData",
        "label": "USD Coin",
    },
]

REQUIRED_HEADERS = {
    "date_iso",
    "time_berlin",
    "symbol",
    "price_usd",
}


def parse_price_value(value: str) -> str | float | None:
    """
    Wandelt Preisfelder optional in echte Zahlen um.

    Leere Felder werden im Zahlenmodus zu None.
    In JavaScript erscheint dies als null.
    """
    value = value.strip()

    if not PRICE_VALUES_AS_NUMBERS:
        return value

    if value == "":
        return None

    normalized = value.replace(",", ".")

    try:
        number = float(normalized)
    except ValueError as exc:
        raise ValueError(
            f'Ungültiger Preiswert: "{value}"'
        ) from exc

    if not math.isfinite(number):
        raise ValueError(
            f'Ungültiger Preiswert: "{value}"'
        )

    return number


def csv_to_date_object(
    csv_file_path: Path,
) -> dict[str, dict[str, Any]]:
    """
    Liest eine Semikolon-getrennte CSV-Datei ein.

    Ergebnis:

    {
        "2026-06-13": {
            "date_iso": "2026-06-13",
            "time_berlin": "02:00",
            "symbol": "BTC",
            "price_eur": "...",
            "price_usd": "..."
        }
    }
    """

    if not csv_file_path.exists():
        raise FileNotFoundError(
            f"CSV-Datei nicht gefunden: {csv_file_path}"
        )

    if csv_file_path.stat().st_size == 0:
        raise ValueError(
            f"CSV-Datei ist leer: {csv_file_path}"
        )

    data: dict[str, dict[str, Any]] = {}

    # utf-8-sig entfernt bei Bedarf automatisch ein UTF-8-BOM.
    with csv_file_path.open(
        "r",
        encoding="utf-8-sig",
        newline="",
    ) as file:

        reader = csv.DictReader(
            file,
            delimiter=";",
        )

        if reader.fieldnames is None:
            raise ValueError(
                f"CSV-Datei besitzt keinen Header: {csv_file_path}"
            )

        headers = [
            header.strip()
            for header in reader.fieldnames
        ]

        reader.fieldnames = headers

        missing_headers = REQUIRED_HEADERS.difference(headers)

        if missing_headers:
            missing = ", ".join(
                sorted(missing_headers)
            )

            raise ValueError(
                f"Pflichtspalten fehlen in "
                f"{csv_file_path.name}: {missing}"
            )

        for line_number, raw_row in enumerate(
            reader,
            start=2,
        ):
            if raw_row is None:
                continue

            # Komplett leere Zeilen ignorieren.
            if all(
                value is None or str(value).strip() == ""
                for value in raw_row.values()
            ):
                continue

            row: dict[str, Any] = {}

            for header in headers:
                raw_value = raw_row.get(header)

                if raw_value is None:
                    raise ValueError(
                        f"Fehlender Wert in "
                        f"{csv_file_path.name}, "
                        f"Zeile {line_number}, "
                        f"Spalte {header}"
                    )

                value = raw_value.strip()

                if header.startswith("price_"):
                    row[header] = parse_price_value(value)
                else:
                    row[header] = value

            date_iso = str(row["date_iso"])

            try:
                datetime.strptime(
                    date_iso,
                    "%Y-%m-%d",
                )
            except ValueError as exc:
                raise ValueError(
                    f'Ungültiges date_iso in '
                    f'{csv_file_path.name}, '
                    f'Zeile {line_number}: '
                    f'"{date_iso}"'
                ) from exc

            # Bei einem doppelten Datum gewinnt
            # die zuletzt gelesene Zeile.
            data[date_iso] = row

    if not data:
        raise ValueError(
            f"CSV enthält keine Datenzeilen: {csv_file_path}"
        )

    # Chronologisch nach Datum sortieren.
    return dict(
        sorted(
            data.items(),
            key=lambda item: item[0],
        )
    )


def build_js_content(
    variable_name: str,
    data: dict[str, dict[str, Any]],
    source_filename: str,
) -> str:

    generated_at = (
        datetime.now(timezone.utc)
        .isoformat()
        .replace("+00:00", "Z")
    )

    json_content = json.dumps(
        data,
        ensure_ascii=False,
        indent=2,
        allow_nan=False,
    )

    return (
        f"// Auto-generiert am {generated_at}\n"
        f"// Quelle: data/price_lists/{source_filename}\n"
        "// Nicht manuell bearbeiten – "
        "stattdessen CSV neu konvertieren.\n\n"

        f"const {variable_name} = {json_content};\n\n"

        "if (typeof globalThis !== 'undefined') {\n"
        f"  globalThis.{variable_name} = {variable_name};\n"
        "}\n\n"

        "if (typeof module !== 'undefined' "
        "&& module.exports) {\n"
        f"  module.exports = {variable_name};\n"
        "}\n"
    )


def write_text_safely(
    output_path: Path,
    content: str,
) -> None:
    """
    Schreibt zuerst in eine temporäre Datei.

    So bleibt bei einem Fehler nicht versehentlich
    eine unvollständige JS-Datei zurück.
    """

    temp_path = output_path.with_suffix(
        output_path.suffix + ".tmp"
    )

    try:
        temp_path.write_text(
            content,
            encoding="utf-8",
            newline="\n",
        )

        temp_path.replace(output_path)

    finally:
        if temp_path.exists():
            temp_path.unlink()


def convert_csv_to_js() -> int:
    OUTPUT_DIR.mkdir(
        parents=True,
        exist_ok=True,
    )

    converted = 0
    skipped = 0
    failed = 0

    for file_config in FILES:
        input_path = (
            INPUT_DIR / file_config["input"]
        )

        output_path = (
            OUTPUT_DIR / file_config["output"]
        )

        if not input_path.exists():
            print(
                f"⚠️  {file_config['label']}: "
                f"{file_config['input']} fehlt – "
                "übersprungen"
            )

            skipped += 1
            continue

        try:
            data = csv_to_date_object(
                input_path
            )

            js_content = build_js_content(
                variable_name=file_config["variable"],
                data=data,
                source_filename=file_config["input"],
            )

            write_text_safely(
                output_path,
                js_content,
            )

            relative_output = output_path.relative_to(
                BASE_DIR
            )

            print(
                f"✅ {file_config['label']}: "
                f"{len(data)} Tage → "
                f"{relative_output}"
            )

            converted += 1

        except Exception as exc:
            print(
                f"❌ {file_config['label']}: {exc}"
            )

            failed += 1

    print(
        "\nFertig: "
        f"{converted} konvertiert, "
        f"{skipped} übersprungen, "
        f"{failed} Fehler."
    )

    if failed > 0 or converted == 0:
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(
        convert_csv_to_js()
    )

File: convert-price-lists-to-js.js
'use strict';

/*
 * Konvertiert die von fill_price_lists_paprika.py erzeugten CSV-Dateien
 * aus data/price_lists/ in einzelne JavaScript-Dateien unter src/scripts/.
 *
 * Verwendung:
 *   node csv_to_js.js
 *
 * Eingabedateien:
 *   data/price_lists/bitcoin_prices.csv
 *   data/price_lists/gomining_prices.csv
 *   data/price_lists/ethereum_prices.csv
 *   data/price_lists/bnb_prices.csv
 *   data/price_lists/solana_prices.csv
 *   data/price_lists/toncoin_prices.csv
 *   data/price_lists/usdt_prices.csv
 *   data/price_lists/usdc_prices.csv
 *
 * Ausgabedateien:
 *   src/scripts/price-data-btc.js
 *   src/scripts/price-data-gmt.js
 *   src/scripts/price-data-eth.js
 *   src/scripts/price-data-bnb.js
 *   src/scripts/price-data-sol.js
 *   src/scripts/price-data-ton.js
 *   src/scripts/price-data-usdt.js
 *   src/scripts/price-data-usdc.js
 */

const fs = require('fs');
const path = require('path');

const INPUT_DIR = path.join(__dirname, 'data', 'price_lists');
const OUTPUT_DIR = path.join(__dirname, 'src', 'scripts');

/*
 * false = Preise bleiben Strings mit deutschem Dezimalkomma,
 *         zum Beispiel "63915,48".
 *
 * true  = price_*-Felder werden zu echten JavaScript-Zahlen,
 *         zum Beispiel 63915.48.
 *
 * Für Berechnungen ist true normalerweise besser.
 * Für maximale Kompatibilität mit deinem alten Format bleibt es hier false.
 */
const PRICE_VALUES_AS_NUMBERS = true;

const FILES = [
  {
    input: 'bitcoin_prices.csv',
    output: 'price-data-btc.js',
    variable: 'btcPriceData',
    label: 'Bitcoin',
  },
  {
    input: 'gomining_prices.csv',
    output: 'price-data-gmt.js',
    variable: 'gmtPriceData',
    label: 'GoMining',
  },
  {
    input: 'ethereum_prices.csv',
    output: 'price-data-eth.js',
    variable: 'ethPriceData',
    label: 'Ethereum',
  },
  {
    input: 'bnb_prices.csv',
    output: 'price-data-bnb.js',
    variable: 'bnbPriceData',
    label: 'BNB',
  },
  {
    input: 'solana_prices.csv',
    output: 'price-data-sol.js',
    variable: 'solPriceData',
    label: 'Solana',
  },
  {
    input: 'toncoin_prices.csv',
    output: 'price-data-ton.js',
    variable: 'tonPriceData',
    label: 'Toncoin',
  },
  {
    input: 'usdt_prices.csv',
    output: 'price-data-usdt.js',
    variable: 'usdtPriceData',
    label: 'Tether',
  },
  {
    input: 'usdc_prices.csv',
    output: 'price-data-usdc.js',
    variable: 'usdcPriceData',
    label: 'USD Coin',
  },
];

function normalizeLineEndings(text) {
  return text
    .replace(/^\uFEFF/, '')
    .replace(/\r\n?/g, '\n');
}

function parsePriceValue(value) {
  if (!PRICE_VALUES_AS_NUMBERS) {
    return value;
  }

  if (value === '') {
    return null;
  }

  const number = Number(value.replace(',', '.'));

  if (!Number.isFinite(number)) {
    throw new Error(`Ungültiger Preiswert: "${value}"`);
  }

  return number;
}

function csvToDateObject(csvFilePath) {
  if (!fs.existsSync(csvFilePath)) {
    throw new Error(`CSV-Datei nicht gefunden: ${csvFilePath}`);
  }

  const csvContent = normalizeLineEndings(
    fs.readFileSync(csvFilePath, 'utf8')
  ).trim();

  if (!csvContent) {
    throw new Error(`CSV-Datei ist leer: ${csvFilePath}`);
  }

  const lines = csvContent
    .split('\n')
    .map(line => line.trim())
    .filter(Boolean);

  if (lines.length < 2) {
    throw new Error(`CSV enthält keine Datenzeilen: ${csvFilePath}`);
  }

  const headers = lines[0]
    .split(';')
    .map(header => header.trim());

  const requiredHeaders = [
    'date_iso',
    'time_berlin',
    'symbol',
    'price_usd',
  ];

  for (const requiredHeader of requiredHeaders) {
    if (!headers.includes(requiredHeader)) {
      throw new Error(
        `Pflichtspalte "${requiredHeader}" fehlt in ${csvFilePath}`
      );
    }
  }

  const data = {};

  for (let lineIndex = 1; lineIndex < lines.length; lineIndex += 1) {
    const values = lines[lineIndex]
      .split(';')
      .map(value => value.trim());

    if (values.length !== headers.length) {
      throw new Error(
        `Spaltenanzahl stimmt nicht in ${path.basename(csvFilePath)}, ` +
        `Zeile ${lineIndex + 1}: ` +
        `erwartet ${headers.length}, erhalten ${values.length}`
      );
    }

    const row = {};

    headers.forEach((header, columnIndex) => {
      const value = values[columnIndex] ?? '';

      row[header] = header.startsWith('price_')
        ? parsePriceValue(value)
        : value;
    });

    const date = row.date_iso;

    if (!/^\d{4}-\d{2}-\d{2}$/.test(date)) {
      throw new Error(
        `Ungültiges date_iso in ${path.basename(csvFilePath)}, ` +
        `Zeile ${lineIndex + 1}: "${date}"`
      );
    }

    /*
     * Falls ein Datum doppelt vorhanden ist,
     * gewinnt die zuletzt gelesene Zeile.
     */
    data[date] = row;
  }

  /*
   * Garantiert chronologische Reihenfolge
   * im erzeugten JavaScript-Objekt.
   */
  return Object.fromEntries(
    Object.entries(data).sort(
      ([dateA], [dateB]) => dateA.localeCompare(dateB)
    )
  );
}

function buildJsContent(variableName, data, sourceFilename) {
  const generatedAt = new Date().toISOString();

  return (
    `// Auto-generiert am ${generatedAt}\n` +
    `// Quelle: data/price_lists/${sourceFilename}\n` +
    `// Nicht manuell bearbeiten – stattdessen CSV neu konvertieren.\n\n` +

    `const ${variableName} = ${JSON.stringify(data, null, 2)};\n\n` +

    `if (typeof globalThis !== 'undefined') {\n` +
    `  globalThis.${variableName} = ${variableName};\n` +
    `}\n\n` +

    `if (typeof module !== 'undefined' && module.exports) {\n` +
    `  module.exports = ${variableName};\n` +
    `}\n`
  );
}

function convertCsvToJs() {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  let converted = 0;
  let skipped = 0;
  let failed = 0;

  for (const file of FILES) {
    const inputPath = path.join(INPUT_DIR, file.input);
    const outputPath = path.join(OUTPUT_DIR, file.output);

    if (!fs.existsSync(inputPath)) {
      console.warn(
        `⚠️  ${file.label}: ${file.input} fehlt – übersprungen`
      );

      skipped += 1;
      continue;
    }

    try {
      const data = csvToDateObject(inputPath);

      const jsContent = buildJsContent(
        file.variable,
        data,
        file.input
      );

      fs.writeFileSync(
        outputPath,
        jsContent,
        'utf8'
      );

      console.log(
        `✅ ${file.label}: ` +
        `${Object.keys(data).length} Tage → ` +
        `${path.relative(__dirname, outputPath)}`
      );

      converted += 1;
    } catch (error) {
      console.error(
        `❌ ${file.label}: ${error.message}`
      );

      failed += 1;
    }
  }

  console.log(
    `\nFertig: ` +
    `${converted} konvertiert, ` +
    `${skipped} übersprungen, ` +
    `${failed} Fehler.`
  );

  if (failed > 0 || converted === 0) {
    process.exitCode = 1;
  }
}

convertCsvToJs();